compiler/hash-typecheck/src/operations/arrays.rs

```diff
@@ -0,0 +1,377 @@
+            ArrayTerm::Repeated(_, repeat) => {
+                let term = self.normalise_node_fully(repeat)?;
+                let Some(length) = try_use_term_as_integer_lit::<_, usize>(self, term) else {
+                    return stuck_normalising();
+                };
+                Ok(Some(length))
+            }
+        }
+    }
+}
+
+impl<E: TcEnv> Operations<ArrayTerm> for Tc<'_, E> {
+    type TyNode = TyId;
+    type Node = TermId;
+
+    fn check(
+        &self,
+        array_term: &mut ArrayTerm,
+        annotation_ty: Self::TyNode,
+        _: Self::Node,
+    ) -> TcResult<()> {
+        self.normalise_and_check_ty(annotation_ty)?;
+
+        let array_len_origin = array_term.length_origin();
+        let (inner_ty, array_len) = self
+            .use_ty_as_array_ty(annotation_ty)?
+            .unwrap_or_else(|| (Ty::hole(array_len_origin.inferred()), None));
+
+        // Now unify that the terms that are specified in the array match the
+        // annotation type.
+        let inferred_len_term = match *array_term {
+            ArrayTerm::Normal(elements) => {
+                self.check_unified_term_list(elements, inner_ty)?;
+                create_term_from_usize_lit(self.target(), elements.len(), array_len_origin)
+            }
+            ArrayTerm::Repeated(term, repeat) => {
+                self.check_node(term, inner_ty)?;
+                self.check_node(repeat, usize_ty(array_len_origin))?;
+                repeat
+            }
+        };
+
+        // Ensure the array lengths match if given
+        if let Some(len) = array_len {
+            if !self.nodes_are_equal(len, inferred_len_term) {
+                return Err(TcError::MismatchingArrayLengths {
+                    expected_len: len,
+                    got_len: inferred_len_term,
+                });
+            }
+        }
+
+        // Either create  a default type, or apply a substitution to the annotation
+        // type.
+        //
+        // - If the array kind is "repeated", the default annotation that we use is an
+        //   array of the specified length.
+        //
+        // - Otherwise, we just default to a list type.
+        if let Ty::Hole(_) = *annotation_ty.value() {
+            let default_annotation = match array_term {
+                ArrayTerm::Normal(_) => list_ty(inner_ty, NodeOrigin::Expected),
+                ArrayTerm::Repeated(_, repeat) => array_ty(inner_ty, *repeat, NodeOrigin::Expected),
+            };
+
+            self.check_by_unify(default_annotation, annotation_ty)?
+        };
+
+        Ok(())
+    }
+
+    fn try_normalise(
+        &self,
+        _item: ArrayTerm,
+        _item_node: Self::Node,
+    ) -> NormaliseResult<ControlFlow<TermId>> {
+        normalise_nested()
+    }
+
+    fn unify(
+        &self,
+        _src: &mut ArrayTerm,
+        _target: &mut ArrayTerm,
+        src_node: Self::Node,
+        _target_node: Self::Node,
+    ) -> TcResult<()> {
+        // @@Todo
```

Comment:
is this meant to be a todo?

Reply:
Yes, because the unification is not implemented (currently it just blocks)